### ISSUE TITLE
[Feature request] Using a single Torch binary for all the model related content

### ISSUE THREAD

**Issue by erogol** [coqui-ai/TTS#1035]
**🚀 Feature Description**

We consider using a single torch binary file as the model release format as opposed to the current zip files compressing all the other files used by a model at inference. 

It'd help to keep things more compact and easy to manage by the ModelManager such as avoiding setting the relative paths of the dependencies (normalization stats, speaker IDs, etc.) in the config.json.

**Solution**

We'll use a single `model.pth` that includes all the data dependencies and the model itself. Then you can access the content as python attributes like 

```python
import torch
chkp = torch.load("model.pth.tar")
speaker_ids = chkp['speaker_ids']
speaker_encoder = chkp['speaker_encoder']
...
model = chkp['model]'
```

So let us know what you think about this proposed change under this issue.

**Comment by thorstenMueller** [comment]
Sounding good. So just one .pth.tar file instead of
- checkpoint<nnnn>.pth.tar
- config.json
- stats.npy
- ...

Any adjustments needed to keep existing models compatible?

**Comment by erogol** [comment]
Yep only a .pth file (same file but we don't use .tar any more) 

Yes we 'll update all the models the same way

**Comment by jreus** [comment]
Sounds good to me. :-)   ...will there also be a simple way to update our own self-trained existing models?

**Comment by WeberJulian** [comment]
How would fine-tuning work? Say I want to FT an existing model with another dataset. I need to load the 1gb checkpoint in python do `chkp['config']['datasets'][0] = BaseDatasetConfig(name="vctk", meta_file_train="", path="../VCTK/")` then save a 1Gb copy of it if I don't want to overwrite the previous one, and then continue with that checkpoint? Instead of editing the config.json

How about having this compact format for inference only. But training behavior is unchanged and you need a script to build the binary (this can be done each time the best_model is updated during training for example). That way you have clean auto sufficient binaries for inference, and for FT, you can use an 'unpack script' on existing binary to make it easy to experiment with config.

**Comment by erogol** [comment]
```
How would fine-tuning work? Say I want to FT an existing model with another dataset. I need to load the 1gb checkpoint in python do chkp['config']['datasets'][0] = BaseDatasetConfig(name="vctk", meta_file_train="", path="../VCTK/") then save a 1Gb copy of it if I don't want to overwrite the previous one, and then continue with that checkpoint? Instead of editing the config.json
```

How about if we keep config.json separate? 


```
How about having this compact format for inference only.
```

Nothing stays inference only 😄. Also, it creates a decoupling that we need to manage, maintain and book. IMO a single format that works everywhere (train, inference, fine-tune) is a more convenient and reliable approach.

**Comment by WeberJulian** [comment]
> How about if we keep config.json separate?

Oh yeah that would make tinkering much simpler, seems like a good traid-off to me.

**Comment by WeberJulian** [comment]
Maybe we could store model_args in the checkpoint so that you can load the model without the config for inference. And you can overwrite that by specifying a config_file.

**Comment by vince62s** [comment]
As a newbie to coqui what made me confused is the mix of things that are part of chkp['config']
so 
1) I fully agree to embed speakers and languages in the pth.tar file
2) I would advocate (seems to be similar to @WeberJulian) to take out the "datasets" from chkp['config']['dataset'] and leave it ONLY in the config.json (same for test_sentences)

**Comment by erogol** [comment]
currently, chkp['config'] is the exact copy of config.json and it is not really used anywhere in the code. Our code currently relies on config.json.  Sorry for the confusion. 

But adding speaker embeddings and IDs to the checkpoint makes sense. Thanks for the feedback.
